### ISSUE TITLE
expand the usages of 'using' for cleaner code

### ISSUE THREAD

**Issue by beuthkaiser** [nim-lang/Nim#12873]
### Summary
'using' is a keyword targeted at avoiding redundant repetitions. Currently it can only be used top-level, but there are common use-cases on proc-level where it would be beneficial to explicitly define shorthands or a context. The desired functionality can already be done using macros, so the issue is more about this being a default language feature.

### Description
Redundancies in procs are especially common when modeling classes, because member functions are primarily about operating on the data of it's objects. This is not an issue in traditional OOP-languages, because they bind methods to objects and then use the context of the class to infer which members were meant. I'd argue that binding methods to objects is a bad abstraction that introduces unnecessary awkwardness. Instead the separation of data and procs coupled with the proc-calling-syntax that nim provides is the way to go, because it has all the advantages without the drawbacks. The verbosity this produces trough repeated 'self references', as well as unnecessary repetitions in general, can be avoided by providing some syntactic sugar to declare shorthands or context.

I encounter this issue a lot with simple procs, where the repetition just seems silly:
```
proc turn(game :var NimGame) =
  echo "There ",are_or_is(game.left), game.left, " left."
  if game.heros_turn:
    game.act(game.hero)
  else:
    game.act(game.villain)
  game.heros_turn = not game.heros_turn
```
with equations, which imho should read as closely as possible to their text book counterpart:
```
# annoying to type, read and debug:
proc set_a(lune : var Lune, new_a : float) =
  lune.a = new_a
  lune.area_abc = sqrt((lune.a+lune.b+lune.c) * (lune.b+lune.c-lune.a) * (lune.c+lune.a-lune.b) * (lune.a+lune.b-lune.c)) / 4
  lune.area_small_lune = 2*lune.area_abc + lune.a*lune.a * arccos( (pow(lune.b,2)-pow(lune.a,2)-pow(lune.c,2)) / (2*lune.a*lune.c) ) - pow(lune.b,2) * arccos( (pow(lune.b,2)+pow(lune.c,2)-pow(lune.a,2)) / (2*lune.b*lune.c) )
  lune.area_section = PI * pow(lune.a,2) - lune.area_small_lune
  lune.area_large_lune = PI * pow(lune.b,2) - lune.area_section  
```
and with complexer algorithms on class object data, where I find the micro-managing of self references to be a waste of cognitive resources, getting in the way of solving the actual problem.

### Proposed solution

Scope-level 'using':
```
proc set_a(lune : var Lune, new_a : float) =
  using
    a : lune.a
    b : lune.b
    c : lune.c       
  a = new_a
  lune.area_abc = sqrt((a+b+c) * (b+c-a) * (c+a-b) * (a+b-c)) / 4
  lune.area_small_lune = 2*area_abc + a*a * arccos( (pow(b,2)-pow(a,2)-pow(c,2)) / (2*a*c) ) - pow(b,2) * arccos( (pow(b,2)+pow(c,2)-pow(a,2)) / (2*b*c) )
  lune.area_section = PI * pow(a,2) - lune.area_small_lune
  lune.area_large_lune = PI * pow(b,2) - lune.area_section  
```
Scoped 'using':
```
proc turn(game :var NimGame) =
  using game:
    echo "There ",are_or_is(left), left, " left."
    if heros_turn:
      hero.act()
    else:
      villain.act()
    heros_turn = not heros_turn
```
I consider the scope-level version a win-win. This would also allow for verbose/descriptive parameter names exposed to the proc user while using abbreviations inside the proc.

I am sure the scoped version can be abused if one tries hard enough. If you are dealing with similarily named local and global identifiers it probably is inappropriate, because you can get ambiguity and accidental shadowing. If it is overused, it could make it hard to follow where the variables come from and hurt readability instead of help it. 
I do not think this would actually be a big deal in practice, it is mostly useful for OOP, and then global identifiers should not be that relevant anyways. And when dealing with local identifiers it is only an issue when using similar names and/or a lot of local identifiers. And then it is easy enough not to use the feature or use the more explicit scope-level version. But 'member functions' that take no input or only take a little input are not exactly rare creatures. Also being scoped, by using small scopes in places where there is a high amount of repetition, imho it is still very clear where the identifier comes from, it's just a nicer way to write the same thing. To each his own, in my mind at least the benefits outweigh the cost by far.

### Alternatives
This functionaliy can be achieved with macros (see '[with](https://github.com/zevv/with/blob/master/with.nim)' ). I argue it should be a standard language feature, because:
- it has common use cases, especially when doing OOP
- language consistency: there is already an 'using' keyword on top-level, so providing the same functionality on lower levels seems logical
  - standardization: avoid people writing different macros to solve the same issue, make it usable for tutorials etc.
  - reuse an existing keyword
  - make 'using' less obscure of a keyword
 
### Additional Information
similar issue to python's 'self', I have measured 'self' to make up something like 10% of OOP-python code. Some people like it, some hate it. I remember one guy telling about looking for another language to teach his kid how to code, because his son was getting frustrated of having to write self over and over again.

**Comment by mratsim** [comment]
Transferring to RFC

**Comment by demotomohiro** [comment]
> language consistency: there is already an 'using' keyword on top-level, so providing the same functionality on lower levels seems logical

Using statement in Nim and `using` you are talking about looks different thing.
Existing Using statement take a parameter name in left side of a colon and type name in right side.
https://nim-lang.github.io/Nim/manual.html#statements-and-expressions-using-statement

Allowing using statement inside proc and applying it to nested procs is more consistent and less obscure, isn't it?

For example:

```Nim
type
  Foo = object
    f: int
  Bar = object
    f: float

proc myproc() =
  using
    x: Foo

  proc innerProc(x) =
    echo x.f

  var a: Foo
  innerProc(a)

proc myproc2() =
  using
    x: Boo

  proc innerProc(x) =
    echo x.f

  var b: Bar
  innerProc(b)

myproc()
myproc2()
```

> standardization: avoid people writing different macros to solve the same issue, make it usable for tutorials etc.

Putting `[with](https://github.com/zevv/with)` macro to `sugar` module in standard library doesn't solve these problem?
Do people really writing different macros to solve this issue instead of using `with` nimble module?

**Comment by dawkot** [comment]
```nim
using
    a : lune.a
    b : lune.b
    c : lune.c  
```
The same can be achieved by writing:
```nim
template a: untyped = lune.a
template b: untyped = lune.b
template c: untyped = lune.c
```

**Comment by mratsim** [comment]
This intersects somewhat with the `alias` sugar proposals https://github.com/nim-lang/Nim/pull/11822 https://github.com/nim-lang/Nim/pull/11992

**Comment by alehander92** [comment]
i did write my own `update` macro so yes people do
